test/sslapitest.c

```diff
@@ -557,10 +557,19 @@ static int test_keylog_no_master_key(void)
 static int verify_retry_cb(X509_STORE_CTX *ctx, void *arg)
 {
     int res = X509_verify_cert(ctx);
+    int idx = SSL_get_ex_data_X509_STORE_CTX_idx();
+    SSL *ssl;
+
+    /* this should not happen but check anyway */
+    if (idx < 0
+        || (ssl = X509_STORE_CTX_get_ex_data(ctx, idx)) == NULL)
+        return -1;
```

Comment:
> The only valid return values from the callback are 0 or 1

Should this new callback code return `-1` then, especially since this code is also used in the _documentation_ as an example of a callback implementation (AFAICT)?

If you decide to polish this, consider making SSL_get_ex_data_X509_STORE_CTX_idx() and X509_STORE_CTX_get_ex_data() calls conditional on `res` being zero and the error being `X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY`. While the primary result would be the same, localizing those two calls would clarify the intent and help those developers who use such test (and also documentation!) code to bootstrap their applications. Besides, it might protect the error stack from irrelevant entries in hypothetical cases when one of those two calls fail outside the `X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY` scope.

Reply:
> Should this new callback code return `-1` then, especially since this code is also used in the _documentation_ as an example of a callback implementation (AFAICT)?

Ah, good point.